test/gtest/ucs/test_memtype_cache.cc

```diff
@@ -47,6 +62,162 @@ class test_memtype_cache : public ucs::test_with_param<ucs_memory_type_t> {
+        if (allocated_buffers != NULL) {
+            if (buf == allocated_buffers->back()) {
+                allocated_buffers->pop_back();
+            } else if (buf == allocated_buffers->front()) {
+                allocated_buffers->pop_front();
+            } else {
+                UCS_TEST_ABORT("provided mem_bufer must be equal to either front() or back()");
+            }
+        }
+
+        released_ptrs->push_back(std::make_pair(buf->ptr(), buf->size()));
+
+        delete buf;
+    }
+
+    void test_ptrs_released(std::vector<std::pair<void*, size_t> > *released_ptrs) const {
+        while (!released_ptrs->empty()) {
+            void *ptr   = released_ptrs->back().first;
+            size_t size = released_ptrs->back().second;
+
+            test_ptr_released(ptr, size);
+            test_ptr_not_found(ptr, size);
+
+            released_ptrs->pop_back();
+        }
+    }
+
+    void release_buffers(std::deque<mem_buffer*> *allocated_buffers, bool from_head) const {
+        std::vector<std::pair<void*, size_t> > released_ptrs;
+
+        while (!allocated_buffers->empty()) {
+            release_mem_buffer(from_head ?
+                               allocated_buffers->front() :
+                               allocated_buffers->back(),
+                               &released_ptrs, allocated_buffers);
+        }
+
+        test_ptrs_released(&released_ptrs);
+    }
+
+    size_t get_test_step(size_t portions = 64) const {
+        return (RUNNING_ON_VALGRIND ?
+                (ucs_get_page_size() / 2 - 1) :
+                (ucs_get_page_size() / portions));
+    }
+
+    void test_memtype_cache_alloc_diff_mem_types(bool from_head,
+                                                 bool keep_buffers,
+                                                 bool same_size_buffers) {
+        const size_t step       = get_test_step();
+        const size_t inner_step = (same_size_buffers ?
+                                   ucs_get_page_size() : step);
+        std::vector<std::pair<void*, size_t> > released_ptrs;
+        std::deque<mem_buffer*> allocated_buffers;
+
+        if (!keep_buffers && !same_size_buffers && from_head) {
+            /* this test doesn't support push_front */
+            return;
+        }
+
+        const std::vector<ucs_memory_type_t> supported_mem_types =
+            mem_buffer::supported_mem_types();    
+
+        /* The tests try to allocate two buffers with different memory types */
+        for (std::vector<ucs_memory_type_t>::const_iterator iter =
+                 supported_mem_types.begin();
+             iter != supported_mem_types.end(); ++iter) {
+            for (size_t i = 1; i <= ucs_get_page_size(); i += step) {
+                mem_buffer *buf1 = allocate_mem_buffer(i, GetParam(),
+                                                       &allocated_buffers,
+                                                       from_head, 0);
```

Comment:
how does it matter if the allocated buffer is added to head or the tail of the deque?

Reply:
this changes the sequence of freeing memory

Reply:
but release_buffers() also accepts "from_head" parameter which tells which order to release
so why is it needed during allocation?
seems it has no effect really..

Reply:
removed it, now we are just using `std::vector` and insert into (remove from) vector's tail